cmd/http/listener.go

```diff
@@ -73,20 +73,58 @@ func isHTTPMethod(s string) bool {
 	return false
 }
 
-func getResourceHost(bufConn *BufConn, maxHeaderBytes, methodLen int) (resource string, host string, err error) {
+func getPlainText(bufConn *BufConn) (bool, error) {
+	defer bufConn.setReadTimeout()
+
+	if bufConn.canSetReadDeadline() {
+		// Set deadline such that we close the connection quickly
+		// of no data was received from the Peek()
+		bufConn.SetReadDeadline(time.Now().UTC().Add(time.Second * 3))
```

Comment:
The actual problem is reading N bytes from the connection whereas the connection has < N bytes and the connection is not (half) closed.

I think we could handle `Host:` header and/or `\r\n` detection by limiting up to bytesRead from `Read()`.  `EOF` detection can be avoided.

Reply:
This PR already resolves all of those cases by reading one byte at a time by subsequently moving forward upto maxHeadedBytes.  This implementation is similar to textproto.MimeHeader() reader. 

We cannot do a Read() call explicitly simply because we cannot read underlying reader before giving it to Go http reader.

Reply:
I thought of reading one byte at a time in previous PR.  IMO that's not an efficient way to do.  My suggestion is as per my comment in @poornas PR i.e. remove bufio usage and add our own `Peek()`.

Reply:
I am not able to see any performance problems on initial tests,  because reading 1 byte is only first attempt, then subsequently bufio actually reads quickly to a higher value. So most operations are already buffered values.

Adding our own Peek is pretty complex IMO, bufio is neat for this reason.

Reply:
Below is the one does peek support.  Choice is yours.

```diff

diff --git a/cmd/http/bufconn.go b/cmd/http/bufconn.go
index c57a6833..2cd0aca3 100644
--- a/cmd/http/bufconn.go
+++ b/cmd/http/bufconn.go
@@ -17,7 +17,6 @@
 package http
 
 import (
-	"bufio"
 	"net"
 	"net/http"
 	"time"
@@ -26,7 +25,7 @@ import (
 // BufConn - is a generic stream-oriented network connection supporting buffered reader and read/write timeout.
 type BufConn struct {
 	QuirkConn
-	bufReader              *bufio.Reader            // buffered reader wraps reader in net.Conn.
+	peekBuf                []byte
 	readTimeout            time.Duration            // sets the read timeout in the connection.
 	writeTimeout           time.Duration            // sets the write timeout in the connection.
 	request                *http.Request            // HTTP request information.
@@ -73,13 +72,40 @@ func (c *BufConn) RemoveTimeout() {
 // Peek - returns the next n bytes without advancing the reader.  It just wraps bufio.Reader.Peek().
 func (c *BufConn) Peek(n int) ([]byte, error) {
 	c.setReadTimeout()
-	return c.bufReader.Peek(n)
+
+	if c.peekBuf != nil {
+		if len(c.peekBuf) > n {
+			return c.peekBuf[:n], nil
+		}
+
+		n -= len(c.peekBuf)
+	}
+
+	buf := make([]byte, n)
+	bytesRead, err := c.QuirkConn.Read(buf)
+	if c.peekBuf != nil {
+		c.peekBuf = append(c.peekBuf, buf[:bytesRead]...)
+	} else {
+		c.peekBuf = buf[:bytesRead]
+	}
+
+	return c.peekBuf, err
 }
 
 // Read - reads data from the connection using wrapped buffered reader.
 func (c *BufConn) Read(b []byte) (n int, err error) {
 	c.setReadTimeout()
-	n, err = c.bufReader.Read(b)
+
+	if c.peekBuf != nil {
+		n = copy(b, c.peekBuf)
+		c.peekBuf = c.peekBuf[:n]
+		if len(c.peekBuf) == 0 {
+			c.peekBuf = nil
+		}
+	} else {
+		n, err = c.QuirkConn.Read(b)
+	}
+
 	if err == nil && c.updateBytesReadFunc != nil {
 		c.updateBytesReadFunc(c.request, n)
 	}
@@ -102,7 +128,6 @@ func (c *BufConn) Write(b []byte) (n int, err error) {
 func newBufConn(c net.Conn, readTimeout, writeTimeout time.Duration) *BufConn {
 	return &BufConn{
 		QuirkConn:    QuirkConn{Conn: c},
-		bufReader:    bufio.NewReader(c),
 		readTimeout:  readTimeout,
 		writeTimeout: writeTimeout,
 	}
```

Reply:
Just tested only this patch it doesn't work without the changes in the PR @balamurugana 

```
curl -v localhost:9000
* Rebuilt URL to: localhost:9000/
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 9000 (#0)
> GET / HTTP/1.1
> Host: localhost:9000
> User-Agent: curl/7.58.0
> Accept: */*
> 
< HTTP/1.1 400 Bad Request
< Content-Type: text/plain; charset=utf-8
< Connection: close
< 
* Recv failure: Connection reset by peer
* stopped the pause stream!
* Closing connection 0
curl: (56) Recv failure: Connection reset by peer
400 Bad Request
```

`mc ls` everything just hangs. 

```
mc ls --debug myminio
mc: <DEBUG> GET / HTTP/1.1
Host: localhost:9000
User-Agent: Minio (linux; amd64) minio-go/v6.0.6 mc/DEVELOPMENT.GOGET
Authorization: AWS4-HMAC-SHA256 Credential=minio/20180913/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=**REDACTED**
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20180913T074917Z
Accept-Encoding: gzip

mc: <DEBUG> GET / HTTP/1.1
Host: localhost:9000
User-Agent: Minio (linux; amd64) minio-go/v6.0.6 mc/DEVELOPMENT.GOGET
Authorization: AWS4-HMAC-SHA256 Credential=minio/20180913/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=**REDACTED**
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20180913T074917Z
Accept-Encoding: gzip
```

I will stick with this PR implementation for now. It is well tested ..